src/client/metric.rs

```diff
@@ -0,0 +1,496 @@
+pub struct Unit {
+    /*
+        pmapi representation of a unit. below, 31 refers to MSB
+        bits 31 - 28 : space dim   (signed)
+             27 - 24 : time dim    (signed)
+             23 - 20 : count dim   (signed)
+             19 - 16 : space scale (unsigned)
+             15 - 12 : time scale  (unsigned)
+             11 - 8  : count scale (unsigned)
+              7 - 0  : zero pad
+    */
+    pmapi_repr: u32
+}
+
+macro_rules! check_dim (
+    ($dim:expr) => (
+        if $dim > 7 || $dim < -8 {
+            return Err(format!("Unit dimension {} is out of range [-8, 7]", $dim))
+        }
+    )
+);
+
+impl Unit {
+    /// Returns an empty unit with all dimensions set to `0`
+    /// and all scales set to an undefined variant
+    pub fn new() -> Self {
+        Unit {
+            pmapi_repr: 0
+        }
+    }
+
+    /// Modifies and returns the unit with given space scale and dimension
+    pub fn space(mut self, scale: Space, dim: i8) -> Result<Self, String> {
+        check_dim!(dim);
+        self.pmapi_repr |= (scale as u32) << 16;
+        self.pmapi_repr |= ((dim as u32) & ((1 << 4) - 1)) << 28;
+        Ok(self)
+    }
+
+    /// Modifies and returns the unit with given time scale and dimension
+    pub fn time(mut self, time: Time, dim: i8) -> Result<Self, String> {
+        check_dim!(dim);
+        self.pmapi_repr |= (time as u32) << 12;
+        self.pmapi_repr |= ((dim as u32) & ((1 << 4) - 1)) << 24;
+        Ok(self)
+    }
+
+    /// Modifies and returns the unit with given count scale and dimension
+    pub fn count(mut self, count: Count, dim: i8) -> Result<Self, String> {
+        check_dim!(dim);
+        self.pmapi_repr |= (count as u32) << 8;
+        self.pmapi_repr |= ((dim as u32) & ((1 << 4) - 1)) << 20;
+        Ok(self)
+    }
+}
+
+#[derive(Copy, Clone)]
+/// Semantic for a Metric
+pub enum Semantics {
+    /// Counter
+    Counter  = 1,
+    /// Instant
+    Instant  = 3,
+    /// Discrete
+    Discrete = 4
+}
+
+/// Singleton metric
+pub struct Metric<T> {
+    name: String,
+    item: u32,
+    sem: Semantics,
+    indom: u32,
+    unit: u32,
+    shorthelp: String,
+    longhelp: String,
+    val: T,
+    mmap_view: MmapViewSync
+}
+
+lazy_static! {
+    static ref SCRATCH_VIEW: MmapViewSync = {
+        Mmap::anonymous(super::STRING_BLOCK_LEN as usize, Protection::ReadWrite).unwrap()
+            .into_view_sync()
+    };
+}
+
+impl<T: MetricType + Clone> Metric<T> {
+    /// Creates a new PCP MMV Metric.
+    ///
+    /// The value type for the metric is determined and fixed
+    /// at compile time.
+    ///
+    /// `item` is the unique metric ID component of the PMID. Only
+    /// the least-significant 10 bits are used.
+    ///
+    /// `name` length should not exceed 63 bytes
+    ///
+    /// `shorthelp_text` length should not exceed 255 bytes
+    ///
+    /// `longhelp_text` length should not exceed 255 bytes
+    pub fn new(
+        name: &str, item: u32, sem: Semantics,
```

Comment:
We should probably set the item id ourselves instead of allowing it in the constructor. As noted in the example, it is passed zero for all metrics, which would be a problem as it is supposed to be unique for all metrics registered at a particular instant for `pmval`. Also, the initial value can probably come right after name instead of so far down the line, as it is also being used to infer the type of the metric